Doc 4286: RPC permission templates (#913)
corda/corda-docs-portal: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md

---
date: '2023-01-10'
menu:
  corda-5-beta:
    identifier: corda-5-beta-rpc-permissions-templates
    weight: 6000
    parent: corda-5-cli-reference
section_menu: corda-5-beta
title: "RPC Permission Templates"
---

By default, when a cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. RPC permission templates enable you to create fine-grained roles for specific actions such as:

* A dedicated role which can create users, roles, permissions and drive all the associations between them.
* A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload).

> * A dedicated role which allows flows to run on this vNode.

Reviewer: ```suggestion
* A dedicated role which allows flows to run on this virtual node.
```
Please use "virtual node" for consistency.

{{< note >}}
These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
{{< /note >}}

> | Role   | Description |

Reviewer 2: The descriptions aren't clear. If I understand correctly, each is a role that grants some permissions to a user? So the descriptions should all start with something like:
"Allows a user to... "
or
"Grants the necessary permissions to..."

| ----------- | ----------- |
| `UserAdminRole` | Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.       |

> | `VNodeCreatorRole` | Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|

Reviewer: ```suggestion
| `VNodeCreatorRole` | Sets all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
```

> | `CordaDeveloperRole` | Users can use VNode reset and Vnode status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|

Reviewer: ```suggestion
| `CordaDeveloperRole` | Users can use virtual node reset and virtual node status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
```

> | `FlowExecutorRole`|  Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.|

Reviewer: ```suggestion
| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command will permit creation of such role for a given virtual node to start new flows and enquire about the status of the running flows.|
```

> | `FlowExecutorRole`|  Once the virtual node is created, a new CLI command will permit creation of such role for a given virtual node to start new flows and enquire about the status of the running flows.|

Reviewer 2: ```suggestion
| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command permits the creation of such a role for a given virtual node to start new flows and enquire about the status of the running flows.|
```
I don't understand this. What is the "new" command? Are you giving permission to the vnode or to a user to run flows on the vnode?

## Querying Permissions via HTTP RPC

> At the moment it is only possible to retrieve a permission if its identifier is known to the caller.

Reviewer: ```suggestion
At the moment, it is only possible to retrieve a permission if its identifier is known to the caller.
```

> However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.

Reviewer: ```suggestion
However, there might be cases when a new role is being set up and existing permissions need to be included. It would be helpful to do a search of the permissions available.
```

> However, there might be cases when a new role is being set up and existing permissions need to be included. It would be helpful to do a search of the permissions available.

Reviewer 2: "It would be helpful to do a search of the permissions available."

Helpful for what? How is the search done?

{{< note >}}

> Being unable to do this creates potentially duplicated permissions.

Reviewer 2: unable to do what?

{{< /note >}}

> ## Check Permissions When Starting Flows

Reviewer: ```suggestion
## Checking Permissions When Starting Flows
```

Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.

> These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.

Reviewer 2: ```suggestion
These checks should be performed against the RBAC sub-system before passing the start request to a FlowWorker.
```

| `CordaDeveloperRole` | Users can use VNode reset and Vnode status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
| `FlowExecutorRole`|  Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.|

## Querying Permissions via HTTP RPC

At the moment it is only possible to retrieve a permission if its identifier is known to the caller.
However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.

{{< note >}}
Being unable to do this creates potentially duplicated permissions.
{{< /note >}}

## Check Permissions When Starting Flows

Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.

These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.

> ## RPC Permission Templates

Reviewer 3: @Niamh25 - I still think we should remove this test plan section from the documentation.

Author: @vkolomeyko Apologies, I didn't realise the entire section was to go. It has been removed now!

`CordaDeveloperRole` - Users can do the following:

1. VNode reset
2. Vnode status update

This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.

`FlowExecutorRole` - Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.

## Querying Permissions via HTTP RPC

At the moment it is only possible to retrieve a permission if its identifier is known to the caller.
However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.

{{< note >}}
Being unable to do this creates potentially duplicated permissions.
{{< /note >}}

## Check Permissions When Starting Flows

Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.

These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.

## RPC Permission Templates

> This describes a draft manual QA script that can be executed to independently verify that functionality is working as expected.

Reviewer 3: I still think that we should remove "manual QA script" from this document.

5. Use Swagger UI, as "UserAdmin" to create user" CordaDeveloper" and assign the "CordaDeveloperRole" to it.
6. Build a CPI with some usable flows.
7. Log on as "vNodeCreator" and upload CPI and create a vNode.
8. Create a 'FlowExecutorRole" role for running flows for a vNode using the CLI tool:

`>java -jar corda-cli.jar initial-rbac flow-executor -t <https://localhost:8888> -u UserAdmin -p UserAdmin -v 9B02C806787D`

{{< note >}}
This is done using "UserAdmin" and not as all mighty "admin".
After creation role name should be: "FlowExecutorRole-9B02C806787D"
{{< /note >}}

9. Use Swagger UI, as "UserAdmin" to create user "FlowExecutor" and assign "FlowExecutorRole-9B02C806787D" to it.
10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
11. Execute some negative testing scenarios, e.g.:
• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.

> ## Create and Assign Multiple Permissions to a Role

Reviewer 3: This seems unrelated to "RPC Permission Templates" and I propose to remove this section.
OpenAPI call specification along with parameters will be documented along with the rest of the OpenAPI calls.

---
date: '2023-01-10'
menu:
  corda-5-beta:
    identifier: corda-5-beta-rpc-permissions-templates
    weight: 6000
    parent: corda-5-cli-reference
section_menu: corda-5-beta

> title: "RPC permission templates"

Reviewer 2: ```suggestion
title: "RPC Permission Templates"
```

---

> ## Create Common RPC Permission Templates

Reviewer 2: ```suggestion

```

> Currently, when the C5 Cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. The idea is to create fine-grained roles for specific actions such as:

Reviewer 2: ```suggestion
By default, when a cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. RPC permission templates enable you to create fine-grained roles for specific actions such as:
```

> • A dedicated role which can create users, roles, permissions and drive all the associations between them;

Reviewer 3: Probably needs an extra empty line or else it is rendered like so:
![image](https://user-images.githubusercontent.com/31008341/212139655-ad881721-9d7d-492f-9846-805742228ef9.png)

```suggestion
• A dedicated role which can create users, roles, permissions and drive all the associations between them;

```

> • A dedicated role which can create users, roles, permissions and drive all the associations between them;

Reviewer 2: ```suggestion
* A dedicated role which can create users, roles, permissions and drive all the associations between them.
```

> • A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload);

Reviewer 2: ```suggestion
* A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload).
```

> • A dedicated role which allows flows to run on this vNode.

Reviewer 3: ```suggestion
• A dedicated role which allows flows to run on a previously created vNode.
```

> • A dedicated role which allows flows to run on this vNode.

Reviewer 2: ```suggestion
* A dedicated role which allows flows to run on this vNode.
```

{{< note >}}

> These roles and permissions enable certain common operations. Creation of users and associating such users to the roles should be still done manually by an admin user.

Reviewer 2: ```suggestion
These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
```

{{< /note >}}

> UserAdminRole- Creates roles, permissions, and controls all associations between the user, roles, and permissions. Create this role at cluster bootstrap by the admin user. There is no direct write into DB like previously done with the admin user.

Reviewer 3: ```suggestion
`UserAdminRole` - Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.
```

> UserAdminRole- Creates roles, permissions, and controls all associations between the user, roles, and permissions. Create this role at cluster bootstrap by the admin user. There is no direct write into DB like previously done with the admin user.

Reviewer 2: I can't quite follow this an the following lines. Seems like it should be a table or subsections maybe?

> VNodeCreatorRole- Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.

Reviewer 3: ```suggestion
`VNodeCreatorRole` - Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.
```

> CordaDeveloperRole- Users can do the following:

Reviewer 3: ```suggestion
`CordaDeveloperRole` - Users can do the following:
```

1. VNode reset
2. Vnode status update

This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.

> FlowExecutorRole- Once the vNode is created, a new CLI command will permit creation of such role for a given vNode.

Reviewer 3: ```suggestion
`FlowExecutorRole` - Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.
```

> ## Ability to query permissions via HTTP RPC

Reviewer 2: ```suggestion
## Querying Permissions via HTTP RPC
```

At the moment it is only possible to retrieve a permission if its identifier is known to the caller.
However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.

{{< note >}}
Being unable to do this creates potentially duplicated permissions.
{{< /note >}}

## Implement Permissions Check When Starting a Particular Flow

Currently, the check is made to assert whether a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.

These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.

> ## RPC Permission Templates: Test Plan Draft

Reviewer 3: I wonder if we need to include "Test Plan Draft" into potentially end-user facing document?
Feels like it is more of an internal script which our QA team been following.

5. Use Swagger UI, as "UserAdmin" to create user" CordaDeveloper" and assign the "CordaDeveloperRole" to it.
6. Build a CPI with some usable flows.
7. Log on as "vNodeCreator" and upload CPI and create a vNode.
8. Create a 'FlowExecutorRole" role for running flows for a vNode using the CLI tool:

`>java -jar corda-cli.jar initial-rbac flow-executor -t <https://localhost:8888> -u UserAdmin -p UserAdmin -v 9B02C806787D`

{{< note >}}
This is done using "UserAdmin" and not as all mighty "admin".
After creation role name should be: "FlowExecutorRole-9B02C806787D"
{{< /note >}}

9. Use Swagger UI, as "UserAdmin" to create user "FlowExecutor" and assign "FlowExecutorRole-9B02C806787D" to it.
10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
11. Execute some negative testing scenarios, e.g.:
• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.

> ## Create RBAC Operation to Create and Assign Multiple Permissions to a Role

Reviewer 2: Can we make this heading shorter?

> A dedicated HTTP RPC call which will permit the following:

Reviewer 2: ```suggestion
A dedicated HTTP RPC call which permits the following:
```

This line needs more info I think?

> • Create multiple permissions

Reviewer 3: ```suggestion
• Create multiple permissions

```

> • Create multiple permissions

Reviewer 2: ```suggestion
* Create multiple permissions
```

> • Assign them all to the existing roles in a single call

Reviewer 2: ```suggestion
* Assign them all to the existing roles in a single call
```

all what?

> ## Allow the vNodeld to Pass when Performing RBAC

Reviewer 3: I wonder if we should delete this section completely?
